packages/hardhat-vyper/src/index.ts

```diff
@@ -249,14 +254,24 @@ subtask(TASK_COMPILE_VYPER)
         ({ version }) => version
       );
 
+      const versionsToSettings = Object.fromEntries(
+        config.vyper.compilers.map(({ version, settings }) => [
+          version,
+          settings,
+        ])
+      );
+
       const versionGroups: Record<string, ResolvedFile[]> = {};
       const unmatchedFiles: ResolvedFile[] = [];
 
       for (const file of resolvedFiles) {
         const hasChanged = vyperFilesCache.hasFileChanged(
           file.absolutePath,
           file.contentHash,
-          { version: file.content.versionPragma }
+          {
+            version: file.content.versionPragma,
+            settings: versionsToSettings[file.content.versionPragma] ?? {},
```

Comment:
I might be missing something here, but this seems conceptually wrong. The keys in `versionsToSettings` are versions, but `versionPragma` is a version _range_.

In terms of observable behavior, what you want is this:

1. Some file has version pragma `^0.3.0`
2. That file gets compiled with version X (which satisfies that range)
3. When compiling again, if a) the version that will be used to compile that file changes, or b) the same version is used but the settings changed, then you re-compile.